File: 프로그래머스/unrated/250137. ［PCCP 기출문제］ 1번 ／ 붕대 감기/［PCCP 기출문제］ 1번 ／ 붕대 감기.py
def solution(bandage, health, attacks):
    answer = health
    max_time = bandage[0]
    recovery = bandage[1]
    bonus_recovery = bandage[2]
    first_attacked = attacks[0]

    answer += first_attacked[0] - 1 * recovery

    if (first_attacked[0] > max_time):
        answer += bonus_recovery

    if answer >= health:
        answer = health

    answer -= first_attacked[1]

    for i in range(1, len(attacks)):
        previous_time = attacks[i - 1][0]
        attacked_time = attacks[i][0]
        damage = attacks[i][1]

        recovery_hp = attacked_time - previous_time - 1

        answer += recovery_hp * recovery

        if recovery_hp >= max_time:
            answer += (recovery_hp / max_time) * bonus_recovery

        if answer >= health:
            answer = health

        answer -= damage

        if answer <= 0:
            answer = -1
            break

    if answer <= 0:
        answer = -1

    return answer

File: 프로그래머스/1/250137. ［PCCP 기출문제］ 1번 ／ 붕대 감기/［PCCP 기출문제］ 1번 ／ 붕대 감기.java
class Solution {
    public int solution(int[] bandage, int health, int[][] attacks) {
        int answer = health;
        int[] damages = new int[attacks[attacks.length - 1][0]+1];
        for(int i = 0; i<attacks.length; i++) {
            damages[attacks[i][0]] = attacks[i][1];
        }
        
        int max = bandage[0];
        int secondH = bandage[1];
        int addH = bandage[2];
        
        int current = 0;
        
        for(int i = 0; i<damages.length; i++) {
            if (damages[i] == 0) {
                current++;
                answer += secondH;
                if (current == max) {
                    current = 0;
                    answer += addH;
                }
            } else {
                answer -= damages[i];
                if (answer <= 0) {
                    return -1;
                }
                current = 0;
            }
            
            if (answer > health) {
                answer = health;
            }
        }
        
        for(int i = 0; i<damages.length; i++) {
             System.out.print(damages[i] + " ");
        }
        return answer;
    }
}